src/setuptools_pyproject_migration/__init__.py

```diff
@@ -91,8 +91,50 @@ def _generate(self) -> Pyproject:
         if dependencies:
             pyproject["project"]["dependencies"] = dependencies
 
+        entrypoints = self._generate_entrypoints()
+
+        # GUI scripts and console scripts go separate in dedicated locations.
+        if "console_scripts" in entrypoints:
+            pyproject["project"]["scripts"] = entrypoints.pop("console_scripts")
+
+        if "gui_scripts" in entrypoints:
+            pyproject["project"]["gui-scripts"] = entrypoints.pop("gui_scripts")
+
+        # Anything left over gets put in entry-points
+        if entrypoints:
+            pyproject["project"]["entry-points"] = entrypoints
+
         return pyproject
 
+    def _parse_entrypoint(self, entrypoint: str) -> Dict[str, str]:
+        """
+        Extract the entry-point and name from the string.
+        """
+        # Format: 'hello-world = timmins:hello_world'
+        if "=" not in entrypoint:
+            raise ValueError("Entry point %r is not of the form 'name = target'" % entrypoint)
+
+        (name, target) = entrypoint.split("=", 1)
+        return {name.strip(): target.strip()}
+
+    def _generate_entrypoints(self) -> Dict[str, Dict[str, str]]:
+        """
+        Dump the entry-points for the project of the given type, if any.
+        """
+        entrypoints: Dict[str, Dict[str, str]] = {}
+
+        if self.distribution.entry_points:
+            for eptype, raweps in self.distribution.entry_points.items():
+                scripts: Dict[str, str] = {}
+
+                for epstr in raweps:
+                    scripts.update(self._parse_entrypoint(epstr))
+
+                if scripts:
+                    entrypoints[eptype] = scripts
```

Comment:
I think it'd be nice to make the method take `entry_points` as an argument, instead of accessing `self.distribution` directly. That way it would be more clear what the inputs are and easier to reason about the function, e.g. it'll be obvious that it's not accessing random other information from the `Distribution` object.  That would also be reflected in testability, because you could then make it a static method and write some doctests, like with `_parse_entrypoint()`.

Also, if you make `_parse_entrypoint()` return a two-element tuple instead of a dict, you could construct the inner dict a little more efficiently:

```suggestion
            for eptype, raweps in self.distribution.entry_points.items():
                scripts: Dict[str, str] = dict(map(self._parse_entrypoint, raweps))

                if scripts:
                    entrypoints[eptype] = scripts
```

And in fact, I think it would be reasonable to propagate empty lists to empty dicts - that is, if there is a key that exists in `self.distribution.entry_points` with an empty list as its corresponding value, we could just produce an empty list rather than omitting that key. I'm not sure if `setuptools` would ever actually produce an empty list of entry points in that context, but if it does, it presumably means that someone intentionally put a header with no values in `setup.cfg` and if so they would probably want it preserved in `pyproject.toml`. (This would be a good thing to add a test for) Anyway, if you don't worry about empty lists then it simplifies the implementation further:

```suggestion
            for eptype, raweps in self.distribution.entry_points.items():
                entrypoints[eptype] = dict(map(self._parse_entrypoint, raweps))
```

And in fact the entire thing _could_ be condensed to a one-liner, in which case you wouldn't even need to pre-declare `entrypoints`, the whole function body could just be this:

```python
        if not self.distribution.entry_points:
            return {}

        return {
            eptype: dict(map(self._parse_entrypoint, raweps))
            for eptype, raweps in self.distribution.entry_points.items()
        }
```

(or whatever formatting black comes up with) Though this might be going too far down the road of conciseness :stuck_out_tongue: Personally I would stick with the second suggestion.

Reply:
Yeah, I do use list comprehensions to generate things, I keep forgetting the same trick works with `dict`s.

One thing in the back of my mind is whether there's any possibility of `setuptools` throwing us an empty list of strings as an entry point list.

The suggestions here will yield an empty `dict` as the value rather than omitting it completely.  However, the risk of this is probably quite small -- if I go the second option, we can always add a `if not raweps: continue` to the loop later. :-)

Reply:
> One thing in the back of my mind is whether there's any possibility of setuptools throwing us an empty list of strings as an entry point list.

Yeah fair point, but I think if it causes a problem we can fix it like you said. And it's possible the type checker would catch that anyway.